Make simplify_exprt::get_values file-local

tautschnig opened the pull request:
This procedure is used in exactly one method, and does not use any class
members. Thus, move it to the file that method resides in. Also remove
the now-unnecessary forall_ macro and replace the previous use by ranged
for.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Make simplify_exprt::get_values file-local

This procedure is used in exactly one method, and does not use any class
members. Thus, move it to the file that method resides in. Also remove
the now-unnecessary forall_ macro and replace the previous use by ranged
for.

martin-cs reviewed (approved):
I believe the changes are as described and are advantageous, so approve...

.... but ...

Would you consider a more radical change.  There are some other code quality issues here.

*. The name is kind of ... generic, can we have something more specific?
*. No documentation of what it should do.
*. Why it does what it does is ... not entirely clear to me.  I guess you are supposed to iterate over the children of an arithmetic node and pull out all the constants used (for constant folding?) but ... then why the recursion on `ID_if`.
*. It does naive structural induction rather than a visit.  I suspect that for 99.999% of use cases that is fine but if we have another lot of auto-generated code with `() ? :` nested 700 deep, it could get very slow and this kind of issue is a pain to find.
*. The review notes on github say that it isn't covered.  If that isn't just a UI or CI race condition issue then ... maybe this is not needed at all?  Which might actually be more worrying.
*. The "`to_integer` returns false on success" threw me for a moment, maybe a comment?
*.

kroening commented:
BTW, this kind of change is clearly a pattern.
I would like to question whether this is a good idea -- the coupling between the function and the class is lost. It only remains via the fact that they are in the same file, which is very indirect.
If the goal is to document that the function/method does not read or write the members of the class, then it could be made a static member function.

peterschrammel commented:
I've fought too many fights on this topic so that I've probably become less religious about it being a private static class member. The only rather minor advantage of the file-local variant I recognise is that there is no member declaration cluttering the header file and requiring maintenance.

tautschnig commented:
> BTW, this kind of change is clearly a pattern.
> I would like to question whether this is a good idea -- the coupling between the function and the class is lost. It only remains via the fact that they are in the same file, which is very indirect.
> If the goal is to document that the function/method does not read or write the members of the class, then it could be made a static member function.

In addition to what @peterschrammel said: I'm generally happy either way, except I'd expect a class member to be sufficiently well documented or otherwise clear in its purpose. The functions I'm making file-local in these PRs really can only be used in exactly the way they are being used right now, and even using them from other class members runs the risk of violating invariants. If that is resolved and the functions become more generally usable, then I'm all for making them static members.

tautschnig committed:
Cleanup get_values (renaming it to collect_constants) and test it

We can substantially simplify the logic here as there is no need to
compare all values when working with sorted containers.

tautschnig commented:
> I believe the changes are as described and are advantageous, so approve...
> 
> .... but ...
> 
> Would you consider a more radical change. There are some other code quality issues here.
> 

Thank you for raising this. I've now added a second commit that mostly covers your concerns, I believe:

> *. The name is kind of ... generic, can we have something more specific?

Renamed to `collect_constants` and ...

> *. No documentation of what it should do.

... documented.

> *. Why it does what it does is ... not entirely clear to me. I guess you are supposed to iterate over the children of an arithmetic node and pull out all the constants used (for constant folding?) but ... then why the recursion on `ID_if`.

Hopefully that has now become clearer.

> *. It does naive structural induction rather than a visit. I suspect that for 99.999% of use cases that is fine but if we have another lot of auto-generated code with `() ? :` nested 700 deep, it could get very slow and this kind of issue is a pain to find.

Ack, but not actually changed. It is possible that we have such expressions, but I am not sure I can get the restriction only-work-on-if-statements right with an iterator.

> *. The review notes on github say that it isn't covered. If that isn't just a UI or CI race condition issue then ... maybe this is not needed at all? Which might actually be more worrying.

I've now added a test that exercises this code. One aspect of the code, I believe, is actually dead (and I have added a comment saying as much) for other simplification rules already catch this beforehand.

> *. The "`to_integer` returns false on success" threw me for a moment, maybe a comment?
> *.

Replaced by `numeric_cast` (somewhere I've got a commit lying around that cleans out a lot of the `to_integer` uses already).

martin-cs reviewed (approved):
Much better -- thanks for making a second pass.

tautschnig committed:
Remove redundant simplification rule

The simplification of if-expressions over constants within an
(in)equality is already covered elsewhere, as confirmed both by the test
added in the preceding commit (which still succeeds) and test coverage.

tautschnig commented:
> Much better -- thanks for making a second pass.

I've made a third pass: the test confirmed that this simplification rule doesn't even fully trigger, making the code either dead or useless. I have thus added a further commit that completely removes the code.
